src/components/scrollManager.js

```diff
@@ -0,0 +1,246 @@
+        cancelAnimationFrame(scrollTimer);
+        scrollTimer = undefined;
+    }
+
+    function scrollTo(x, y, smooth) {
+        //console.log('scrollTo (' + x + ', ' + y + ', ' + smooth + ')');
+
+        if (smooth !== true) {
+            resetScroll();
+            origScrollTo(x, y);
+            return true;
+        }
+
+        if (scrollTimer) {
+            scrollTarget.x = trend(window.scrollX, scrollTarget.x, x);
+            scrollTarget.y = trend(window.scrollY, scrollTarget.y, y);
+        } else {
+            scrollTarget.x = x;
+            scrollTarget.y = y;
+        }
+
+        var scrollMaxX = Math.max(0, Math.max(document.body.scrollWidth, document.body.offsetWidth, document.documentElement.clientWidth, document.documentElement.scrollWidth, document.documentElement.offsetWidth) - window.innerWidth);
+        var scrollMaxY = Math.max(0, Math.max(document.body.scrollHeight, document.body.offsetHeight, document.documentElement.clientHeight, document.documentElement.scrollHeight, document.documentElement.offsetHeight) - window.innerHeight);
+
+        scrollTarget.x = clamp(0, scrollTarget.x, scrollMaxX);
+        scrollTarget.y = clamp(0, scrollTarget.y, scrollMaxY);
+
+        if (scrollTarget.x === window.scrollX && scrollTarget.y === window.scrollY) {
+            //console.log('cancel scroll animation');
+            resetScroll();
+            return true;
+        }
+
+        //console.log('scroll from (' + window.scrollX + ', ' + window.scrollY + ') to (' + scrollTarget.x + ', ' + scrollTarget.y + ')');
+
+        if (!scrollTimer) {
+            var start;
+
+            function scrollAnim(currentTimestamp) {
+                start = start || currentTimestamp;
+
+                var dx = scrollTarget.x - window.scrollX;
+                var dy = scrollTarget.y - window.scrollY;
+                //console.log('> ' + dx + ', ' + dy + ' = ' + window.scrollX + ', ' + window.scrollY);
+
+                if (Math.abs(dx) <= 1 && Math.abs(dy) <= 1) {
+                    //console.log('scrolled');
+                    scrollTimer = undefined;
+                    origScrollTo(scrollTarget.x, scrollTarget.y);
+                    return;
+                }
+
+                var k = Math.min(0.5, (currentTimestamp - start) / scrollTime);
+
+                dx = Math.sign(dx)*Math.ceil(Math.abs(dx*k));
+                dy = Math.sign(dy)*Math.ceil(Math.abs(dy*k));
+
+                //console.log('> ' + k + ' > ' + dx + ', ' + dy);
+                origScrollBy(dx, dy);
+
+                scrollTimer = requestAnimationFrame(scrollAnim);
+            };
+
+            scrollTimer = requestAnimationFrame(scrollAnim);
+        }
+
+        return false;
+    }
+
+    function scrollBy(dx, dy, smooth) {
+        //console.log('scrollBy (' + dx + ', ' + dy + ', ' + smooth + ')');
+        return scrollTo(window.scrollX + dx, window.scrollY + dy, smooth);
+    }
+
+    function scrollIntoView(elem, scrollX, scrollY) {
```

Comment:
a comment explaining what this function should do and its parameters would be nice